### PR TITLE
Add the Queue stat counters to Counter DB

### PR DESCRIPTION
**Pull request by andrii-savka**

### CODE REVIEW

File: syncd/syncd_saiswitch.cpp [hunk 1]
```diff
@@ -500,20 +494,16 @@ void SaiSwitch::removeExistingObject(
...
 
-std::vector<sai_port_stat_t> SaiSwitch::saiGetSupportedCounters() const
+void SaiSwitch::saiInitSupportedPortCounters()
 {
     SWSS_LOG_ENTER();
 
-    auto ports = saiGetPortList();
-
-    if (ports.size() == 0)
+    if (m_port_rid_list.size() == 0)
```
Comment (qiluo-msft) [inline syncd/syncd_saiswitch.cpp]: >m_port_rid_list.size() == 0 [](start = 8, length = 27)

!m_port_rid_list.empty()

File: syncd/syncd_saiswitch.cpp [hunk 2]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("failed to collect number of queues for port RID %s: %s",
+                    sai_serialize_object_id(port_rid).c_str(),
+                    sai_serialize_status(status).c_str());
+            continue;
+        }
+
+        std::vector<sai_object_id_t> queue_list;
+        queue_list.resize(sai_attr.value.u32);
```
Comment (qiluo-msft) [inline syncd/syncd_saiswitch.cpp]: std::vector<sai_object_id_t> queue_list(sai_attr.value.u32).

File: syncd/syncd_saiswitch.cpp [hunk 3]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+
+        m_queue_rid_map[port_rid] = queue_list;
+    }
+}
+
+void SaiSwitch::saiInitSupportedQueueCounters()
+{
+    SWSS_LOG_ENTER();
+
+    if (m_port_rid_list.size() == 0)
```
Comment (qiluo-msft) [inline syncd/syncd_saiswitch.cpp]: >m_port_rid_list.size() == 0 [](start = 8, length = 27)

empty()

File: syncd/syncd_saiswitch.cpp [hunk 4]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+    }
+
+    sai_object_id_t port_rid = m_port_rid_list.at(0);
+
+    if (m_queue_rid_map.find(port_rid) == m_queue_rid_map.end())
+    {
+        SWSS_LOG_THROW("no queues are defined on port");
+    }
+
+    sai_object_id_t queue_rid = m_queue_rid_map.at(port_rid).at(0);
```
Comment (qiluo-msft) [inline syncd/syncd_saiswitch.cpp]: >at(port_rid) [](start = 48, length = 12)

All ready 'find' by previous statement, no need to 'at'.

File: syncd/syncd_saiswitch.cpp [hunk 5]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+    if (m_port_rid_list.size() == 0)
+    {
+        SWSS_LOG_THROW("no ports are defined on switch");
+    }
+
+    sai_object_id_t port_rid = m_port_rid_list.at(0);
+
+    if (m_queue_rid_map.find(port_rid) == m_queue_rid_map.end())
+    {
+        SWSS_LOG_THROW("no queues are defined on port");
```
Comment (qiluo-msft) [inline syncd/syncd_saiswitch.cpp]: Add port_rid into the string?

File: syncd/syncd_saiswitch.cpp [hunk 6]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+                    sai_serialize_object_id(port_rid).c_str(),
+                    sai_serialize_status(status).c_str());
+
+            continue;
+        }
+
+        m_supported_queue_counters.push_back(counter);
+    }
+
+    SWSS_LOG_NOTICE("supported %zu of %d",
```
Comment (qiluo-msft) [inline syncd/syncd_saiswitch.cpp]: >%d [](start = 38, length = 2)

unsigned?

File: syncd/syncd_saiswitch.cpp [hunk 7]
```diff
@@ -100,21 +100,19 @@ std::string SaiSwitch::saiGetHardwareInfo() const
     return std::string(info);
 }
 
-std::vector<sai_object_id_t> SaiSwitch::saiGetPortList() const
+void SaiSwitch::saiInitPortList()
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: saiGetinitPortList

Comment (kcudnik) [reply syncd/syncd_saiswitch.cpp]: also shit method should be const, and return list of ports, if you want to store them inside SaiSwitch object, call this method inside SaiSwitch constructor and then populate m_port_rid_list, and change name to something like m_init_port_list_rids etc

ps. why you need init port list? we agreed thta port number will not change during runtime for example ports will not be splitted/merged

File: syncd/syncd_saiswitch.cpp [hunk 8]
```diff
@@ -128,11 +126,9 @@ std::vector<sai_object_id_t> SaiSwitch::saiGetPortList() const
                 sai_serialize_status(status).c_str());
     }
 
-    portList.resize(attr.value.objlist.count);
-
-    SWSS_LOG_DEBUG("number of ports: %zu", portList.size());
+    m_port_rid_list.resize(attr.value.objlist.count);
 
-    return portList;
+    SWSS_LOG_DEBUG("number of ports: %zu", m_port_rid_list.size());
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: number of init ports ?

File: syncd/syncd_saiswitch.cpp [hunk 9]
```diff
@@ -500,20 +494,16 @@ void SaiSwitch::removeExistingObject(
     }
 }
 
-std::vector<sai_port_stat_t> SaiSwitch::saiGetSupportedCounters() const
+void SaiSwitch::saiInitSupportedPortCounters()
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: NO!. all saiXXX functions are just call to sai, and should be const, and should return queries value, bring that back!

File: syncd/syncd_saiswitch.cpp [hunk 10]
```diff
@@ -599,9 +589,7 @@ void SaiSwitch::collectCounters(
             const std::string &field = sai_serialize_port_stat(m_supported_counters[idx]);
             const std::string &value = std::to_string(counters[idx]);
 
-            swss::FieldValueTuple fvt(field, value);
-
-            values.push_back(fvt);
+            values.push_back( swss::FieldValueTuple(field, value) );
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: dont gibe extra spaces '( swss' this formatting is not used anywhere in this project

File: syncd/syncd_saiswitch.cpp [hunk 11]
```diff
@@ -610,6 +598,58 @@ void SaiSwitch::collectCounters(
...
+{
+    SWSS_LOG_ENTER();
+
+    uint32_t countersSize = (uint32_t)m_supported_queue_counters.size();
+    std::vector<uint64_t> counters;
+    counters.resize(countersSize);
+
+    for (auto &port_rid: m_port_rid_list)
+    {
+        if (m_queue_rid_map.find(port_rid) == m_queue_rid_map.end())
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: reuse iterator instead double looking find() and then at90 on line 619, compare iterator with end()

File: syncd/syncd_saiswitch.cpp [hunk 12]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
     return ita->second;
 }
 
+void SaiSwitch::saiInitQueueMap()
+{
+	SWSS_LOG_ENTER();
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: ident is not equal, should be 4 spaces

File: syncd/syncd_saiswitch.cpp [hunk 13]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+
+    for (auto &port_rid : m_port_rid_list)
+    {
+        sai_attribute_t sai_attr;
+        sai_attr.id = SAI_PORT_ATTR_QOS_NUMBER_OF_QUEUES;
+        sai_status_t status = sai_metadata_sai_port_api->get_port_attribute(port_rid, 1, &sai_attr);
+        if (status != SAI_STATUS_SUCCESS)
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: empty line before if

File: syncd/syncd_saiswitch.cpp [hunk 14]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+            continue;
+        }
+
+        std::vector<sai_object_id_t> queue_list;
+        queue_list.resize(sai_attr.value.u32);
+
+        sai_attr.id = SAI_PORT_ATTR_QOS_QUEUE_LIST;
+        sai_attr.value.objlist.list = queue_list.data();
+        status = sai_metadata_sai_port_api->get_port_attribute(port_rid, 1, &sai_attr);
+        if (status != SAI_STATUS_SUCCESS)
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: empty line before if

File: syncd/syncd_saiswitch.cpp [hunk 15]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+    SWSS_LOG_ENTER();
+
+    if (m_port_rid_list.size() == 0)
+    {
+        SWSS_LOG_THROW("no ports are defined on switch");
+    }
+
+    sai_object_id_t port_rid = m_port_rid_list.at(0);
+
+    if (m_queue_rid_map.find(port_rid) == m_queue_rid_map.end())
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: reuse iterator sincd you are doing double lookup

File: syncd/syncd_saiswitch.cpp [hunk 16]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            const std::string &name = sai_serialize_enum(counter, &sai_metadata_enum_sai_queue_stat_t);
+            SWSS_LOG_DEBUG("counter %s is not supported on port RID %s: %s",
+                    name.c_str(),
+                    sai_serialize_object_id(port_rid).c_str(),
+                    sai_serialize_status(status).c_str());
+
+            continue;
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: sometimes you use empty line before continue, sometimes not, be consistent

File: syncd/syncd_saiswitch.cpp [hunk 17]
```diff
@@ -1193,6 +1233,84 @@ sai_object_id_t SaiSwitch::getDefaultValueForOidAttr(
...
+    for (uint32_t idx = 0; idx < sai_metadata_enum_sai_queue_stat_t.valuescount; ++idx)
+    {
+        sai_queue_stat_t counter = (sai_queue_stat_t)sai_metadata_enum_sai_queue_stat_t.values[idx];
+        uint64_t value;
+        sai_status_t status = sai_metadata_sai_queue_api->get_queue_stats(queue_rid, 1, &counter, &value);
+
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            const std::string &name = sai_serialize_enum(counter, &sai_metadata_enum_sai_queue_stat_t);
+            SWSS_LOG_DEBUG("counter %s is not supported on port RID %s: %s",
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: always addd empty line before LOG line

File: syncd/syncd_saiswitch.cpp [hunk 18]
```diff
@@ -1228,9 +1346,15 @@ SaiSwitch::SaiSwitch(
 
     helperInternalOids();
 
+    saiGetMacAddress(m_default_mac_address);
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.cpp]: just as general rule, don't change order of executed  code!

File: syncd/syncd_saiswitch.h [hunk 19]
```diff
@@ -189,8 +189,14 @@ class SaiSwitch
 
         sai_mac_t m_default_mac_address;
 
+        std::vector<sai_object_id_t> m_port_rid_list;
+
         std::vector<sai_port_stat_t> m_supported_counters;
 
+        std::map<sai_object_id_t, std::vector<sai_object_id_t>>  m_queue_rid_map;
+
+        std::vector<sai_queue_stat_t> m_supported_queue_counters;
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.h]: add brief description to added variables, as you added on line 236

File: syncd/syncd_saiswitch.h [hunk 20]
```diff
@@ -227,11 +233,28 @@ class SaiSwitch
...
 
         std::unordered_map<sai_uint32_t, sai_object_id_t> saiGetHardwareLaneMap() const;
 
-        std::vector<sai_port_stat_t> saiGetSupportedCounters() const;
+        /**
+         * @brief Get the supported port statistic counters for port from SAI.
+         */
+        void saiInitSupportedPortCounters();
+
+
```
Comment (kcudnik) [inline syncd/syncd_saiswitch.h]: doubleempty line

**Comment by andrii-savka** [comment]
The new pull request with review fixes will be added to the master branch.

**Pull request closed without merging**